include/picongpu/plugins/saxs/Saxs.kernel

```diff
@@ -0,0 +1,291 @@
+        using FramePtr = typename ParBox::FramePtr;
+
+        uint32_t const workerIdx = threadIdx.x;
+
+        constexpr int blockSize =
+            pmacc::math::CT::volume<SuperCellSize>::type::value;
+
+        // vectorial part of the integrand in the Jackson formula
+        PMACC_SMEM(acc, r_s, memory::Array<picongpu::float3_64, blockSize>);
+
+        // storage for macro particle weighting
+        PMACC_SMEM(acc, saxsWeighting_s, memory::Array<float_X, blockSize>);
+
+        PMACC_SMEM(acc, np_s, float_64);
+
+        // get extent of guarding super cells (needed to ignore them)
+        DataSpace<simDim> const guardingSuperCells =
+            mapper.getGuardingSuperCells();
+
+        /* number of super cells on GPU per dimension (still including guard
+         * cells) remove both guards from count [later one sided guard needs to
+         * be added again]
+         */
+        DataSpace<simDim> const superCellsCount(
+            mapper.getGridSuperCells() - 2 * guardingSuperCells);
+
+        // get absolute number of relevant super cells
+        int const numSuperCells = superCellsCount.productOfComponents();
+
+        /* go over all super cells on GPU
+         * but ignore all guarding supercells
+         */
+        for (int super_cell_index = 0; super_cell_index <= numSuperCells;
+             ++super_cell_index)
+        {
+            // select SuperCell and add one sided guard again
+            DataSpace<simDim> const superCell =
+                DataSpaceOperations<simDim>::map(
+                    superCellsCount, super_cell_index) +
+                guardingSuperCells;
+
+            // guardingSuperCells remove guarding block
+            DataSpace<simDim> const superCellOffset(
+                globalOffset +
+                ((superCell - guardingSuperCells) * SuperCellSize::toRT()));
+
+            // pointer to frame storing particles
+            FramePtr frame = pb.getLastFrame(superCell);
+
+            // number of particles in current frame
+            lcellId_t particlesInFrame =
+                pb.getSuperCell(superCell).getSizeLastFrame();
+
+            // go to next supercell
+            while (frame.isValid())
+            {
+                /* since a race condition can occur if "continue loop" is
+                 * called, all threads must wait for the selection of a new
+                 * frame until all threads have evaluated "isValid"
+                 */
+                __syncthreads();
+
+                ForEachIdx<IdxConfig<1, numWorker>> onlyMaster{workerIdx};
+
+                /* The Master process (thread 0) in every thread block is in
+                 * charge of loading a frame from
+                 * the current super cell and evaluate the total number of
+                 * particles in this frame.
+                 */
+                onlyMaster([&](uint32_t const, uint32_t const) {
+                    np_s = 0;
+                    nmp[0] += particlesInFrame;
```

Comment:
If we use more than one block to perform the algorithm we need to avoid race conditions. Maybe we should not count particles in this kernel and use the PMacc method for counting macro particles.